pkg/component/server/apiendpointreconciler.go

```diff
@@ -0,0 +1,221 @@
+/*
+Copyright 2020 k0s authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package server
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"reflect"
+	"sort"
+	"sync/atomic"
+	"time"
+
+	config "github.com/k0sproject/k0s/pkg/apis/v1beta1"
+	k0sv1beta1 "github.com/k0sproject/k0s/pkg/apis/v1beta1"
+	k8sutil "github.com/k0sproject/k0s/pkg/kubernetes"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// APIEndpointReconciler is the component to reconcile in-cluster API address endpoint based from externalName
+type APIEndpointReconciler struct {
+	ClusterConfig *config.ClusterConfig
+
+	L *logrus.Entry
+
+	leaderElector     LeaderElector
+	stopCh            chan struct{}
+	kubeClientFactory k8sutil.ClientFactory
+}
+
+// NewEndpointReconciler creates new endpoint reconciler
+func NewEndpointReconciler(c *k0sv1beta1.ClusterConfig, leaderElector LeaderElector, kubeClientFactory k8sutil.ClientFactory) *APIEndpointReconciler {
+	d := atomic.Value{}
+	d.Store(true)
+	return &APIEndpointReconciler{
+		ClusterConfig:     c,
+		leaderElector:     leaderElector,
+		stopCh:            make(chan struct{}),
+		kubeClientFactory: kubeClientFactory,
+		L:                 logrus.WithFields(logrus.Fields{"component": "endpointreconciler"}),
+	}
+}
+
+// Init initializes the APIEndpointReconciler
+func (a *APIEndpointReconciler) Init() error {
+	_, err := net.LookupIP(a.ClusterConfig.Spec.API.ExternalAddress)
+	if err != nil {
+		return fmt.Errorf("cannot resolve api.externalAddress: %w", err)
+	}
+
+	return nil
+}
+
+// Run runs the main loop for reconciling the externalAddress
+func (a *APIEndpointReconciler) Run() error {
+
+	go func() {
+		ticker := time.NewTicker(10 * time.Second)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ticker.C:
+				err := a.reconcileEndpoints()
+				if err != nil {
+					a.L.Warnf("external API address reconciliation failed: %s", err.Error())
+				}
+			case <-a.stopCh:
+				a.L.Info("endpoint reconciler done")
+				return
+			}
+		}
+	}()
+
+	return nil
+}
+
+// Stop stops the reconciler
+func (a *APIEndpointReconciler) Stop() error {
+	close(a.stopCh)
+	return nil
+}
+
+// Healthy dummy implementation
+func (a *APIEndpointReconciler) Healthy() error { return nil }
+
+func (a *APIEndpointReconciler) reconcileEndpoints() error {
+	if !a.leaderElector.IsLeader() {
+		a.L.Debug("we're not the leader, not reconciling api endpoints")
```

Comment:
I think return should be added here, to end function if we are not the leader?

Reply:
oh, very good catch 👍

Reply:
fixed